Addressing problems raised in sphinx_issues #10104

hoangduytranuk opened the pull request:
Subject: Solving problems raised in issues #10104 
- Breaking changes: master
- Feature

### Purpose
- Allowing message locations to be sorted uniquely in alphabetic order. This will remove duplicates locations and enable easier search and identify locations for developers

### Detail
-  remove duplicates locations 
- enabled easier search and identify locations for developers
- 
### Relates
- [Message.locations duplicate unnecessary](https://github.com/sphinx-doc/sphinx/issues/10104)

Commit:
Addressing problems raised in sphinx_issues #10104

AA-Turner commented on sphinx/builders/gettext.py:
```diff
@@ -32,7 +32,7 @@ class Message:
     def __init__(self, text: str, locations: List[Tuple[str, int]], uuids: List[str]):
         self.text = text
-        self.locations = locations
+        self.locations = list(set(locations))
```
I'd use `sorted` here -- `set` doesn't guarantee any sort order at a language level.

```suggestion
        self.locations = sorted({*locations})
```

AA-Turner commented on sphinx/builders/gettext.py:
```diff
@@ -57,7 +57,7 @@ def add(self, msg: str, origin: Union[Element, "MsgOrigin"]) -> None:
     def __iter__(self) -> Generator[Message, None, None]:
         for message in self.messages:
-            positions = [(source, line) for source, line, uuid in self.metadata[message]]
+            positions = [(os.path.relpath(source, start=os.getcwd()), line) for source, line, uuid in self.metadata[message]]
```
This would call `os.getcwd()` for every item in the loop. Does ommitting `start=` still give a usable result? Otherwise you could use cache `os.getcwd()` in a local.

tk0miya replied:
Understandable change.

But this should be moved to the outside of this class because this class does not have appropriate directory information. I think what directory you want is the base directory of the document (a.k.a srcdir). But it seems you're using `os.getcwd()`. But it does not equal to srcdir. It usually equals to the confdir that contains conf.py, but not promised. It sometimes returns a different directory if 3rd party extension changes it.


BTW, please use `sphinx.util.osutil:relpath()` instead of `os.path.relpath()`. It will not be crashed even if `path` and `start` are on different drives on Windows.

hoangduytranuk replied:
Fixed the issue #10104 after the code reviews from Adam Turner and Takeshi Komiya:
- Added the class variable 'builder' to Catalog class, so any instances of Catalog can access the information within the builder, including things such as app, config, srcdir, outdir etc.. see MessageCatalogBuilder for details.
- Initialise this Catalog.builder as the instance of MessageCatalogBuilder is created, ie. in the init()
- As Message class is created, list of locations passing to it will be made singular, using set(), sorted alphabetically for easier searching and identifying origins where the message can be found.
- source dir is used within the __iter__ of the Catalog, using sphinx.util.osutil.relpath, for safer functionality, to work out the relative path for each location entry.

AA-Turner replied:
> Added the class variable 'builder' to Catalog class

Modifying global state like this is a bad idea, please could you refactor?

A

tk0miya replied:
Yes, adding a class variable is not a good idea. `Catalog` class is a data class that represents message catalog. It is not related to the builder at all. So the path manipulation using builder should be called outside of the catalog class. How about processing just before calling `GettextRenderer`?

AA-Turner commented on sphinx/builders/gettext.py:
```diff
@@ -1,5 +1,5 @@
 """The MessageCatalogBuilder class."""
-
+import os
```
Keep a blank line here
```suggestion

import os
```

AA-Turner reviewed (changes requested):
A few comments:

A

hoangduytranuk commented:
Thank you so much for these feedbacks. I will return to fix the code at a later time and post another commit to the pull request. 

Best Regards,
Hoang Tran

> On 23 Apr 2022, at 12:09, Adam Turner ***@***.***> wrote:
> 
> 
> @AA-Turner requested changes on this pull request.
> 
> A few comments:
> 
> A
> 
> In sphinx/builders/gettext.py <https://github.com/sphinx-doc/sphinx/pull/10379#discussion_r856880120>:
> 
> > @@ -32,7 +32,7 @@ class Message:
>      """An entry of translatable message."""
>      def __init__(self, text: str, locations: List[Tuple[str, int]], uuids: List[str]):
>          self.text = text
> -        self.locations = locations
> +        self.locations = list(set(locations))
> I'd use sorted here -- set doesn't guarantee any sort order at a language level.
> 
> ⬇️ Suggested change
> -        self.locations = list(set(locations))
> +        self.locations = sorted({*locations})
> In sphinx/builders/gettext.py <https://github.com/sphinx-doc/sphinx/pull/10379#discussion_r856880517>:
> 
> >          for message in self.messages:
> -            positions = [(source, line) for source, line, uuid in self.metadata[message]]
> +            positions = [(os.path.relpath(source, start=os.getcwd()), line) for source, line, uuid in self.metadata[message]]
> This would call os.getcwd() for every item in the loop. Does ommitting start= still give a usable result? Otherwise you could use cache os.getcwd() in a local.
> 
> In sphinx/builders/gettext.py <https://github.com/sphinx-doc/sphinx/pull/10379#discussion_r856880555>:
> 
> > @@ -1,5 +1,5 @@
>  """The MessageCatalogBuilder class."""
> -
> +import os
> Keep a blank line here
> 
> ⬇️ Suggested change
> -import os
> +
> +import os
> —
> Reply to this email directly, view it on GitHub <https://github.com/sphinx-doc/sphinx/pull/10379#pullrequestreview-950954435>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKIONBYYLLCUQ4CPX5HMNNDVGPK6RANCNFSM5UEIUEYQ>.
> You are receiving this because you authored the thread.
>

hoangduytranuk committed:
Fixed the issue #10104 after the code reviews from Adam Turner and Takeshi Komiya:
- Added the class variable 'builder' to Catalog class, so any instances of Catalog can access the information within the builder, including things such as app, config, srcdir, outdir etc.. see MessageCatalogBuilder for details.
- Initialise this Catalog.builder as the instance of MessageCatalogBuilder is created, ie. in the init()
- As Message class is created, list of locations passing to it will be made singular, using set(), sorted alphabetically for easier searching and identifying origins where the message can be found.
- source dir is used within the __iter__ of the Catalog, using sphinx.util.osutil.relpath, for safer functionality, to work out the relative path for each location entry.

AA-Turner commented on sphinx/builders/gettext.py:
```diff
@@ -32,13 +32,14 @@ class Message:
     def __init__(self, text: str, locations: List[Tuple[str, int]], uuids: List[str]):
         self.text = text
-        self.locations = locations
+        self.locations = list(sorted(list(set(locations))))     #make list unique and sorted alphabetically
```
`sorted` returns a list and can take a set as input

```suggestion
        self.locations = sorted(set(locations))  # sort and de-duplicate the locations list
```
